mesa/experimental/cell_space/grid.py

```diff
@@ -105,7 +106,103 @@ def _connect_single_cell_2d(self, cell: T, offsets: list[tuple[int, int]]) -> No
+        select_random_empty_cell(self): Returns an empty random cell instance from the grid.
+
+
+    Note:
+        This class is not intended for direct use in user models but is currently used by the OrthogonalMooreGrid and OrthogonalVonNeumannGrid.
+    """
+
+    def __init__(
+        self,
+        dimensions: Sequence[int],
+        torus: bool = False,
+        capacity: float | None = None,
+        random: Random | None = None,
+        cell_klass: type[T] = Cell,
+        property_layers: None | PropertyLayer | list[PropertyLayer] = None,
+    ) -> None:
+        super().__init__(
+            dimensions=dimensions,
+            torus=torus,
+            capacity=capacity,
+            random=random,
+            cell_klass=cell_klass,
+        )
+
+        self.properties = {}
+
+        # Handle both single PropertyLayer instance and list of PropertyLayer instances
+        if property_layers:
+            # If a single PropertyLayer is passed, convert it to a list
+            if isinstance(property_layers, PropertyLayer):
+                property_layers = [property_layers]
+
+            for layer in property_layers:
+                self.add_property_layer(layer)
+
+    def add_property_layer(self, property_layer: PropertyLayer):
+        """
+        Adds a new property layer to the grid.
+
+        Args:
+            property_layer (PropertyLayer): The PropertyLayer instance to be added to the grid.
+
+        Raises:
+            ValueError: If a property layer with the same name already exists in the grid.
+            ValueError: If the dimensions of the property layer do not match the grid's dimensions.
+        """
+        if property_layer.name in self.properties:
+            raise ValueError(f"Property layer {property_layer.name} already exists.")
+        if (
+            property_layer.width != self.dimensions[0]
+            or property_layer.height != self.dimensions[1]
+        ):
+            raise ValueError(
+                f"Property layer dimensions {property_layer.width}x{property_layer.height} do not match grid dimensions {self.dimensions[0]}x{self.dimensions[1]}."
+            )
+        self.properties[property_layer.name] = property_layer
+
+    def remove_property_layer(self, property_name: str):
+        """
+        Removes a property layer from the grid by its name.
+
+        Args:
+            property_name (str): The name of the property layer to be removed.
+
+        Raises:
+            ValueError: If a property layer with the given name does not exist in the grid.
+        """
+        if property_name not in self.properties:
+            raise ValueError(f"Property layer {property_name} does not exist.")
+        del self.properties[property_name]
+
+    def select_random_empty_cell(self) -> T:
+        """select random empty cell and add the property to the cell"""
+        cell = super().select_random_empty_cell()
+        for property in self.properties:
```

Comment:
Alright. I thought that being able to access and modify the property layer directly from the cell instance would be a valuable feature. We could implement direct methods for editing and accessing properties within the cell instance."

Reply:
Yes, definitely, but I **guess** what @rht meant was that this link between the cell and the property layer currently is only established once `select_random_empty_cell` is called. It would be better if the link is established immediately when a property layer is added to the space.

Reply:
@Corvince Yes, In that case when we add a property layer, it would necessitate the creation of a cell instance for each cell. Could this potentially lead to unnecessary memory consumption?

Reply:
So, do it the other way around. Let cells look up the properties based on their coordinates.

This does imply that there should be an unambiguous way to map coordinates from cells to the property layer.

Reply:
"Yes. However, considering the inefficiency of both proposed approaches—creating a cell instance for every cell in the grid and adding a property to the cell instance under 'select_random_empty_cell'—I am uncertain about the implementation. One alternative could be refraining from storing the property in the cell upon its instantiation. Instead, we could define a method named 'property(property_name)' within the cell class. This method would return the property value when called upon.

Reply:
> @Corvince Yes, In that case when we add a property layer, it would necessitate the creation of a cell instance for each cell. Could this potentially lead to unnecessary memory consumption?

Cells are already instantiated when a space is created. You just need to iterate all the cells and add the properties exactly how you do it already here in the select_random_cell function. 

Memory consumption should not be an issue, you are just creating some references to the property layers and the actual values. So it will only be a few bytes per cell. 

And I think then we are already good to merge (but we should wait for @ewouth s review, since mainly he created the PropertyGrid)

Reply:
I will try to review it Sunday!